### ISSUE TITLE
False problem reported on `__new__` method of a metaclass

### ISSUE THREAD

**Issue by yaxum62** [microsoft/pyright#2640]
**Describe the bug**
Pyright reports problem when working on first argument of `__new__` method in a metaclass.

**To Reproduce**
```python
from typing import TypeVar, Any
_Type= TypeVar("_Type")

class MetaClass(type):
  def __new__(cls: type[_Type], name: str, bases: tuple[type, ...], namespace: dict[str, Any], **kwds: Any) -> _Type:
    ret = super().__new__(cls, name, bases, namespace, **kw)
    #                     ~~~ Argument of type "Type[_Type@__new__]" cannot be assigned to parameter "cls" of type "Type[_TT@__new__]" in function "__new__"
    return ret
    #      ~~~ Expression of type "_TT@__new__" cannot be assigned to return type "_Type@__new__"
```

**Expected behavior**
Type inference of `__new__` method in metaclass should works the same as normal class. e.g. Following code works without any error:
```python
class Foo(object):
    def __new__(cls: type[_T], *args: object) -> _T:
        return super().__new__(cls)
```

**VS Code extension or command-line**
pylance extension v2021.11.2

**Comment by erictraut** [comment]
The error is because the `_Type` TypeVar is not bound to `type`, as required by the `__new__` method in the `type` class.

The following change will eliminate the error:
```python
_Type = TypeVar("_Type", bound="type")
```

**Comment by yaxum62** [comment]
Ah thanks for the explain, I see what I was missing. To extend this issue a little:

1. Can we include some details in the error? From current message I have no idea why type mismatches. Something like following would be nice: `_Type@__new__ is unbounded but _TT@__new__ is bound to "type"`
1. Figuring out exact requirement for a `TypeVar` is hard (and I bet sometimes it is impossible).  It would be nice if we can make TypeVar requirements lazily checked. e.g. reports error when user invokes `MetaClass.__new__(int, ...)` but not inside `__new__()` function.

**Comment by erictraut** [comment]
I agree that the error message isn't as clear as it could be. There is a bug where the source and destination types were swapped in the error message, which made it more confusing than otherwise. I've fixed that bug. As for reporting that it's bound to "type", that's not feasible. The error is detected deep in the guts of the TypeVar constraint solver, and the source of the bound TypeVar is no longer available.

As for reporting the error in another location, that wouldn't be correct. I think it would also be confusing to report the problem elsewhere. The actual type error occurs at the point where it's currently reported, at the call site to `super().__new__`.
